### PR TITLE
ASM-6510 I want to discover EMC Storage and show it under Resources page and Dashboard

### PR DESCRIPTION
**Pull request by sushilrai**
- Added ping sweep script to detect EMC VNX Service Processor IP
- Added script to fetch facts for EMC VNX to be displayed in ASM UI

### CODE REVIEW

**Comment by gavin-scott** [comment]
@asaenz please review

**Comment by sushilrai** [comment]
@haribabu @gavin-scott  please review and merge the changes. This will enable EMC discovery on ASM Appliance.

File: bin/discovery.rb [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+  opt :server, 'EMC VNX Array', :type => :string, :required => true
+  opt :username, 'EMC VNX username', :type => :string, :required => true
+  opt :password, 'EMC VNX password', :type => :string, :default => ENV['PASSWORD']
+  opt :timeout, 'command timeout', :default => 240
+  opt :output, 'output facts to a file', :type => :string, :required => true
+end
+
+facts = {}
+
+def collect_emc_vnx_facts opts
```
Comment (gavin-scott) [inline bin/discovery.rb]: Missing parentheses around `opts`.

File: bin/discovery.rb [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  facts = {}
+  facts['disk_info'] ||= []
+  disks = xml_doc.xpath('//CLAR:Physicals/CLAR:Disks/CLAR:Disk')
+  disks.each do |disk|
+    disk_info_hash = {}
+    disk.children.each do |disk_attr|
+      next if disk_attr.node_name == "text"
+      disk_info_hash[disk_attr.node_name] = disk_attr.content
+    end
+    facts['disk_info'] << {disk_info_hash.first.last => disk_info_hash}
```
Comment (gavin-scott) [inline bin/discovery.rb]: Same comment applies to a bunch of the structures below, `raid_groups`, `disk_pools`, etc.

Comment (sushilrai) [inline bin/discovery.rb]: Disk Info element contains multiple attributes of each disk

`{"Bus 0 Enclosure 1 Disk 14"=>{"Name"=>"Bus 0 Enclosure 1 Disk 14", "Bus"=>"0", "Enclosure"=>"1", "Slot"=>"14", "Bank"=>"-1", "BankAsInteger"=>"4294967295", "State"=>"1", "DiskState"=>"0", "CapacityInMBs"=>"0", "NumberOfUserSectors"=>"0", "Type"=>"3", "Writes"=>"0", "Reads"=>"0", "BlocksRead"=>"0", "BlocksWritten"=>"0", "WriteRetries"=>"4294967295", "ReadRetries"=>"4294967295", "IdleTicks"=>"0", "BusyTicks"=>"0", "ZeroMark"=>"9223372036854775807", "CurrentSpeed"=>"0", "MaximumSpeed"=>"0", "Replacing"=>"false", "HardReadErrors"=>"0", "HardWriteErrors"=>"0", "SoftReadErrors"=>"0", "SoftWriteErrors"=>"0", "IsHWSpinDownQualified"=>"false", "IsHWSpinDownCapable"=>"false", "IsVaultDrive"=>"false", "PowerSavingState"=>"4", "DiskDriveCategory"=>"15"}}`

File: bin/discovery.rb [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    hba_port_attr = hba_port.children
+    port_hash = {}
+    ['WWN', 'VendorDescription', 'NumberOfSPPorts' ].collect {|x| port_hash[x] = hba_port_attr.at_xpath("//SAN:#{x}").text}
+    hba_ports_info << {port_hash.first.last => port_hash}
+  end
+  hba_facts['hba_ports_info'] = hba_ports_info
+  hba_facts
+end
+
+def collect_inventory opts
```
Comment (hribab) [inline bin/discovery.rb]: Can we do collect_inventory(opts) ? same way while calling it in #26 line

File: bin/discovery.rb (hunk 2 above)
Comment (gavin-scott) [inline bin/discovery.rb]: Yeah, so my question was why we don't return one big hash of e.g. `{"Bus 0 Enclosure 1 Disk 14"=>{...}, "Bus 0 Enclosure 1 Disk 15"=>{...}}`. Currently you return a list of single-element hashes `[{"Bus 0 Enclosure 1 Disk 14"=>{...}}, {"Bus 0 Enclosure 1 Disk 15"=>{...}}]`

File: bin/discovery.rb (hunk 3 above)
Comment (gavin-scott) [inline bin/discovery.rb]: Yes, we should have parentheses.

**Comment by sushilrai** [comment]
@gavin-scott @hribab please review and merge. Without this PR, EMC discovery is broken

**Pull request merged**